File: benchmark/pmrgb/simhash.py
"""64-bit SimHash + Hamming distance for near-duplicate detection.

Mirrors the extension's dedup step (docs/02 §4) so the benchmark can inject and
label realistic near-duplicate pages. Stdlib-only (hashlib).
"""
from __future__ import annotations

import hashlib
import re

_TOKEN = re.compile(r"[a-z0-9]+")


def _token_hash(token: str) -> int:
    h = hashlib.blake2b(token.encode("utf-8"), digest_size=8).digest()
    return int.from_bytes(h, "big")


def simhash64(text: str) -> int:
    tokens = _TOKEN.findall(text.lower())
    if not tokens:
        return 0
    v = [0] * 64
    for tok in tokens:
        hv = _token_hash(tok)
        for i in range(64):
            v[i] += 1 if (hv >> i) & 1 else -1
    out = 0
    for i in range(64):
        if v[i] > 0:
            out |= (1 << i)
    return out


def hamming(a: int, b: int) -> int:
    return bin(a ^ b).count("1")


def is_near_dup(a: str, b: str, threshold: int = 3) -> bool:
    return hamming(simhash64(a), simhash64(b)) <= threshold


def simhash_hex(text: str) -> str:
    return f"0x{simhash64(text):016x}"

File: extension/lib/memory/simhash.ts
/**
 * 64-bit SimHash + Hamming distance for near-duplicate detection (docs/02 §4).
 * Uses BigInt; token hash = 64-bit FNV-1a. Mirrors benchmark/pmrgb/simhash.py.
 */
const MASK = (1n << 64n) - 1n;
const FNV_OFFSET = 1469598103934665603n;
const FNV_PRIME = 1099511628211n;

function fnv1a64(token: string): bigint {
  let h = FNV_OFFSET;
  for (let i = 0; i < token.length; i++) {
    h ^= BigInt(token.charCodeAt(i) & 0xff);
    h = (h * FNV_PRIME) & MASK;
  }
  return h;
}

export function simhash64(text: string): bigint {
  const tokens = text.toLowerCase().match(/[a-z0-9]+/g) ?? [];
  if (!tokens.length) return 0n;
  const v = new Array<number>(64).fill(0);
  for (const tok of tokens) {
    const hv = fnv1a64(tok);
    for (let i = 0; i < 64; i++) v[i] = v[i]! + ((hv >> BigInt(i)) & 1n ? 1 : -1);
  }
  let out = 0n;
  for (let i = 0; i < 64; i++) if (v[i]! > 0) out |= 1n << BigInt(i);
  return out;
}

export function hamming(a: bigint, b: bigint): number {
  let x = a ^ b;
  let c = 0;
  while (x) {
    x &= x - 1n;
    c++;
  }
  return c;
}

export const isNearDup = (a: string, b: string, threshold = 3) =>
  hamming(simhash64(a), simhash64(b)) <= threshold;

export const simhashHex = (text: string) => '0x' + simhash64(text).toString(16).padStart(16, '0');
